src/Workspaces/Core/Portable/Workspace/Host/SyntaxTreeFactory/AbstractSyntaxTreeFactoryService.AbstractRecoverableSyntaxRoot.cs

```diff
@@ -101,18 +102,20 @@ public RecoverableSyntaxRoot<TRoot> WithSyntaxTree(IRecoverableSyntaxTree<TRoot>
                 }
             }
 
-            protected override async Task SaveAsync(TRoot root, CancellationToken cancellationToken)
+            protected override async Task SaveAsync(TRoot root)
             {
                 Contract.ThrowIfFalse(_storage == null); // Cannot save more than once
 
                 // tree will be always held alive in memory, but nodes come and go. serialize nodes to storage
                 using (var stream = SerializableBytes.CreateWritableStream())
                 {
-                    root.SerializeTo(stream, cancellationToken);
+                    root.SerializeTo(stream, CancellationToken.None);
                     stream.Position = 0;
 
-                    _storage = _service.LanguageServices.WorkspaceServices.GetService<ITemporaryStorageService>().CreateTemporaryStreamStorage(cancellationToken);
-                    await _storage.WriteStreamAsync(stream, cancellationToken).ConfigureAwait(false);
+                    var storage = _service.LanguageServices.WorkspaceServices.GetService<ITemporaryStorageService>().CreateTemporaryStreamStorage(CancellationToken.None);
+                    await storage.WriteStreamAsync(stream, CancellationToken.None).ConfigureAwait(false);
+
+                    Interlocked.CompareExchange(ref _storage, storage, null);
```

Comment:
Regardless, please do not use complex concurrency constructs when simple ones will do.  It makes the code much harder to review and much harder to maintain.